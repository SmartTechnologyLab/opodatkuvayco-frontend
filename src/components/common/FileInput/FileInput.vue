<script setup lang="ts">
import { useI18n } from 'vue-i18n'

const { t } = useI18n()

defineEmits(['file-selected'])
</script>

<template>
  <label class="file-input">
    <span class="file-input__download">{{ t('main.calc.file-download') }}</span>
    <input @change="$emit('file-selected')" type="file" style="display: none" />
  </label>
</template>

<style lang="scss" scoped>
$calc-input-background: #384648;

.file-input {
  outline: none;
  background: $calc-input-background;
  border: none;
  height: 60px;
  border-radius: 10px;
  font-size: 20px;
  cursor: pointer;

  display: flex;
  align-items: center;
  justify-content: center;

  &__download {
    text-decoration: underline;
  }
}
</style>
