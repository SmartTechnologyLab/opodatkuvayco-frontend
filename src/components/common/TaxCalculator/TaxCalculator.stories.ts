import TaxCalculator from './TaxCalculator.vue'
import type { Meta, StoryFn } from '@storybook/vue3'

export default {
  title: 'Components/TaxCalculator',
  component: TaxCalculator
} as Meta<typeof TaxCalculator>

const Template: StoryFn<typeof TaxCalculator> = () => ({
  components: { TaxCalculator },
  template: '<TaxCalculator />'
})

export const Default = Template.bind({})
