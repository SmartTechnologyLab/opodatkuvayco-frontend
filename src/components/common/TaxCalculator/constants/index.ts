export const validChars = '0123456789+-/*.'

export const operations: string[] = ['*', '/', '+', '-']

export const values: Array<number | string> = [1, 2, 3, 4, 5, 6, 7, 8, 9, '+/-', 0, '.']
