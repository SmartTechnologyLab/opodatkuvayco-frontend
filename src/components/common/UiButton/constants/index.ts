export enum Icons {
  CHECK = 'pi pi-check',
  USER = 'pi pi-user',
  CROSS = 'pi pi-times',
  DELETE = 'pi pi-trash',
  PLUS = 'pi pi-plus'
}

export enum Severities {
  SECONDARY = 'secondary',
  HELP = 'help',
  DANGER = 'danger'
}
