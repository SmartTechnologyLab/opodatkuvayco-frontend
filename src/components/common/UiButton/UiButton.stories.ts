import UiButton from './UiButton.vue'
import { Icons, Severities } from './constants'
import type { Meta, StoryFn } from '@storybook/vue3'

export default {
  component: UiButton,
  args: {
    hasSlot: false,
    label: 'Button',
    loading: false,
    disabled: false
  },
  argTypes: {
    onClick: { action: 'click' },
    icon: {
      control: 'radio',
      options: [undefined, ...Object.values(Icons)],
      description: 'Choose icon of the button'
    },
    severity: {
      control: 'radio',
      options: Object.values(Severities),
      description: 'Choose icon of the button'
    }
  }
} as Meta<typeof UiButton>

const Template: StoryFn<typeof UiButton> = (args) => ({
  components: { UiButton },
  setup() {
    return { args }
  },
  template: `<div>
    
    <UiButton v-if="args.hasSlot" v-bind="args">{{ $t(args.label) }}</UiButton>
    <UiButton v-else  v-bind="args" />
  </div>
  `
})

export const Default = Template.bind({})
Default.args = {
  icon: undefined,
  severity: Severities.SECONDARY,
  rounded: true,
  outlined: true,
  raised: true
}

export const Colors: StoryFn<typeof UiButton> = (args) => ({
  components: { UiButton },
  setup() {
    return { args, Severities }
  },
  template: `<div style="display: grid; gap: 8px; grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));">
    <UiButton v-for="severity in Severities" :key="severity" v-bind="{ ...args, severity, label: severity}" />
  </div>`
})

export const Ico: StoryFn<typeof UiButton> = (args) => ({
  components: { UiButton },
  setup() {
    return { args, Icons }
  },
  template: `<div style="display: flex; gap: 8px;">
    <UiButton v-for="icon in Icons" :key="icon" v-bind="{ ...args, icon, label: ''}" />
  </div>`
})
