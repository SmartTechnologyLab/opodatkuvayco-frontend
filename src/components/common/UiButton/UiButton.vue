<script setup lang="ts">
import Button from 'primevue/button'
import { computed, useAttrs } from 'vue'
import { Icons, Severities } from './constants'

defineOptions({
  inheritAttrs: false
})

const attrs = useAttrs()

const props = defineProps<{
  icon?: Icons
  severity?: Severities
  rounded?: boolean
  outlined?: boolean
  raised?: boolean
}>()

defineEmits<{
  (e: 'click'): void
}>()

const routerLinkTag = 'router-link'
const linkTag = 'a'

const component = computed(() => {
  if (attrs.to) {
    return routerLinkTag
  }
  if (attrs.href) {
    return linkTag
  }

  return null
})

const componentAttrs = computed(() => {
  const { href, to } = attrs

  if (component.value === 'router-link') {
    return { to }
  }

  return { href }
})
</script>

<template>
  <Component :is="component" v-if="component" v-bind="{ ...componentAttrs }">
    <Button v-bind="{ ...props, ...attrs }" @click="$emit('click')">
      <slot />
    </Button>
  </Component>

  <Button v-bind="{ ...props, ...attrs }" @click="$emit('click')" v-else>
    <slot />
  </Button>
</template>
