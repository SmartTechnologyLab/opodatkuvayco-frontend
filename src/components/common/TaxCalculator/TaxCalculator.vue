<script setup lang="ts">
import FileInput from '@/components/common/FileInput/FileInput.vue'
import { computed, ref } from 'vue'
import { useI18n } from 'vue-i18n'
import UIButton from '@/components/common/UiButton/UiButton.vue'
import { Icons } from '@/components/common/UiButton/constants'
import { operations, validChars, values } from '@/components/common/TaxCalculator/constants/index'

const selectedFiles = ref<File[]>([])
const result = ref('')
const isNegative = ref(false)
const isCalculating = ref(false)

const { t } = useI18n()

const addToResult = (num: string | number) => {
  if (num === '+/-') {
    isNegative.value = true
  } else if (typeof num === 'number' && !isNaN(num)) {
    result.value = result.value.toString()

    result.value += isNegative.value ? '-' + num : num

    isNegative.value = false
  } else {
    result.value += num
  }
}

const reset = () => {
  result.value = ''
  isCalculating.value = false
}

const calculate = () => {
  result.value = eval(result.value) || 0
  isCalculating.value = true
}

const handleFileChange = (event: Event) => {
  const target = event.target as HTMLInputElement
  const files = target?.files

  if (files) {
    selectedFiles.value.push(...files)
  }
}

const handleDeleteFile = (file: File) => {
  selectedFiles.value = selectedFiles.value.filter((selectedFile) => selectedFile !== file)
}

const checkIsValid = (value: string) => validChars.includes(value)

const handleCheckInput = (value: string) => {
  result.value = value
    .split('')
    .filter((char) => checkIsValid(char))
    .join('')
}

const equalBtnType = computed(() => (selectedFiles.value.length && !result.value.length ? 'submit' : 'button'))
</script>

<template>
  <div class="calc">
    <input
      v-if="result.length || isCalculating"
      type="text"
      class="calc__result"
      v-model="result"
      @input="handleCheckInput(result)"
    />

    <FileInput @change="handleFileChange" accept=".json, .xml" v-else />

    <div class="calc__example">
      <span class="calc__example-title" v-if="!selectedFiles.length">
        {{ t('main.calc.file-example') }}
      </span>

      <ul class="calc__files-list" v-else>
        <li class="calc__file" v-for="file in selectedFiles" :key="file.lastModified">
          <span class="calc__file-name">
            {{ file.name }}
          </span>

          <UIButton :icon="Icons.CROSS" @click="handleDeleteFile(file)" class="calc__delete-btn" />
        </li>
      </ul>
      <div class="calc__toggle" />
    </div>

    <div class="calc__container">
      <div class="calc__numbers-grid">
        <UIButton v-for="value in values" :key="value" class="calc__numbers" @click="addToResult(value)">
          {{ value }}
        </UIButton>
      </div>

      <div class="calc__wrapper">
        <div class="calc__operations-grid">
          <UIButton
            v-for="operation in operations"
            @click="addToResult(operation)"
            :key="operation"
            class="calc__numbers"
          >
            {{ operation }}
          </UIButton>
        </div>

        <div class="calc__submit-grid">
          <UIButton class="calc__numbers calc__numbers--equal" @click="reset">C</UIButton>

          <UIButton class="calc__numbers calc__numbers--equal" @click="calculate" :type="equalBtnType"> = </UIButton>
        </div>
      </div>
    </div>
  </div>
</template>

<style scoped lang="scss">
$calc-background: $secondary-background;

$calc-text-color: $secondary-text-color;

$calc-input-background: #384648;

$toggle-background-main: #1c2022;

$toggle-background-secondary: #6691a1;

$special-btn-background: #8cd0d0;

$special-btn-background: #8cd0d0;

.calc {
  padding: 1rem;
  background: $calc-background;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  gap: 1rem;
  width: 100%;
  max-width: 460px;

  &__container {
    display: flex;
    gap: 1rem;
    width: 100%;
  }

  &__example {
    display: flex;
    justify-content: space-between;
    align-items: start;
  }

  &__toggle {
    height: 18px;
    border: 1px solid #628c9c;
    border-radius: 18px;
    width: 4rem;
    background: linear-gradient(to right, $toggle-background-main 70%, $toggle-background-secondary 70%);
  }

  &__example-title {
    text-decoration: underline;
  }

  &__file {
    display: flex;
    align-items: center;
    gap: 0.3rem;
    max-width: 235px;
    overflow: hidden;
  }

  &__files-list {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    gap: 1rem;
    padding: 0;
  }

  &__file-name {
    text-decoration: underline;
    overflow: hidden;
  }

  &__result {
    outline: none;
    background: $calc-input-background;
    border: none;
    height: 60px;
    border-radius: 10px;
    padding-right: 10%;
    font-size: 20px;
    text-align: right;
  }

  &__numbers-grid {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-gap: 1rem 1rem;
    width: 100%;
    height: 15rem;
  }

  &__numbers {
    background: $main-text-color;
    color: $calc-text-color;
    border-radius: 10px;
    font-size: 18px;
    border: none;
    cursor: pointer;
    min-width: 2rem;
    max-width: 4rem;
    height: 2.8rem;
    display: flex;
    align-items: center;
    justify-content: center;

    &--equal {
      min-width: 2rem;
      max-width: 4rem;
      height: 6.9rem;

      &:nth-child(2) {
        background: $special-btn-background;
        color: $main-text-color;
      }
    }
  }

  &__wrapper {
    display: flex;
    flex-direction: column;
    gap: 1rem;
    width: 63%;
  }

  &__operations-grid {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-gap: 1rem;
    height: 100%;
  }

  &__submit-grid {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-gap: 1rem;
  }

  &__result {
    display: flex;
    align-items: center;
    justify-content: center;
  }

  &__delete-btn {
    background: none;
    width: 1.5rem;
  }
}
</style>
