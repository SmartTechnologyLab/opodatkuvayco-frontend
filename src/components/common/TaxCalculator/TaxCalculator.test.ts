import { DOMWrapper, shallowMount, type VueWrapper } from '@vue/test-utils'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import TaxCalculator from '@/components/common/TaxCalculator/TaxCalculator.vue'
import FileInput from '@/components/common/FileInput/FileInput.vue'
import { nextTick } from 'vue'
import UiButton from '../UiButton/UiButton.vue'

const file = new File(['file'], 'example.json', { type: 'application/json' })

const mockDataTransfer = {
  files: [file],
  items: {
    add: vi.fn()
  }
}

describe('TaxCalculator component', () => {
  let wrapper: VueWrapper<InstanceType<typeof TaxCalculator>>

  const findButtonByText = (text: string) => {
    return wrapper.findAllComponents(UiButton).find((button) => button.text() === text)
  }

  beforeEach(() => {
    wrapper = shallowMount(TaxCalculator, {
      global: {
        stubs: {
          UiButton,
          FileInput
        }
      }
    })
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('When result is null or isCalculating is false - FileInput renders', () => {
    const fileInput = wrapper.findComponent(FileInput)
    const resultInput = wrapper.find('input[type="text"]')

    expect(fileInput.exists()).toMatchSnapshot()
    expect(resultInput.exists()).toMatchSnapshot()
  })

  const ONE = '1'
  const TWO = '2'
  const THREE = '3'
  const PLUS = '+'
  const MINUS = '-'
  const MULTIPLY = '*'
  const DIVIDE = '/'
  const EQUAL = '='
  const CLEAR = 'C'
  it('When the button with number is clicked, the result input is mounted into DOM', async () => {
    const numberBtn = findButtonByText(ONE)

    await numberBtn?.trigger('click')

    expect(wrapper.findComponent(FileInput).exists()).toMatchSnapshot()
    expect(wrapper.find('input').exists()).toMatchSnapshot()
  })

  it('When the reset button clicked, the result input is unmounted from DOM and the FileInput renders', async () => {
    const numberBtn = findButtonByText(ONE)
    const resetBtn = findButtonByText(CLEAR)

    await numberBtn?.trigger('click')

    expect(wrapper.findComponent(FileInput).exists()).toBe(false)
    expect(wrapper.find('input').exists()).toBe(true)

    await resetBtn?.trigger('click')

    const resultInput = wrapper.find('input[type="text"]')

    expect(wrapper.findComponent(FileInput).exists()).toMatchSnapshot()
    expect(resultInput.exists()).toMatchSnapshot()
  })

  it('Calctulating the sum correctly', async () => {
    const equalBtn = findButtonByText(EQUAL)
    const plusBtn = findButtonByText(PLUS)
    const oneBtn = findButtonByText(ONE)
    const twoBtn = findButtonByText(TWO)

    await oneBtn?.trigger('click')

    await plusBtn?.trigger('click')

    await twoBtn?.trigger('click')

    await equalBtn?.trigger('click')

    const resultInput = wrapper.find('input[type="text"]')
    expect((resultInput.element as HTMLInputElement).value).toMatchSnapshot()
  })

  it('Calctulating the difference correctly', async () => {
    const equalBtn = findButtonByText(EQUAL)
    const minusBtn = findButtonByText(MINUS)
    const oneBtn = findButtonByText(ONE)
    const twoBtn = findButtonByText(TWO)

    await oneBtn?.trigger('click')

    await minusBtn?.trigger('click')

    await twoBtn?.trigger('click')

    await equalBtn?.trigger('click')

    const resultInput = wrapper.find('input[type="text"]')
    expect((resultInput.element as HTMLInputElement).value).toMatchSnapshot()
  })

  it('Calctulating the multiplication correctly', async () => {
    const equalBtn = findButtonByText(EQUAL)
    const multiplyBtn = findButtonByText(MULTIPLY)
    const twoBtn = findButtonByText(TWO)
    const threeBtn = findButtonByText(THREE)

    await threeBtn?.trigger('click')

    await multiplyBtn?.trigger('click')

    await twoBtn?.trigger('click')

    await equalBtn?.trigger('click')

    const resultInput = wrapper.find('input[type="text"]')
    expect((resultInput.element as HTMLInputElement).value).toMatchSnapshot()
  })

  it('Calctulating the division correctly', async () => {
    const equalBtn = findButtonByText(EQUAL)
    const divisionBtn = findButtonByText(DIVIDE)
    const threeBtn = findButtonByText(THREE)

    await threeBtn?.trigger('click')

    await divisionBtn?.trigger('click')

    await threeBtn?.trigger('click')

    await equalBtn?.trigger('click')

    const resultInput = wrapper.find('input[type="text"]')
    expect((resultInput.element as HTMLInputElement).value).toMatchSnapshot()
  })

  it('Set the number to negative correctly', async () => {
    const equalBtn = findButtonByText(EQUAL)
    const plusBtn = findButtonByText(PLUS)
    const negativeBtn = findButtonByText('+/-')
    const threeBtn = findButtonByText(THREE)

    await threeBtn?.trigger('click')

    await plusBtn?.trigger('click')

    await negativeBtn?.trigger('click')

    await threeBtn?.trigger('click')

    await equalBtn?.trigger('click')

    const resultInput = wrapper.find('input')
    expect((resultInput.element as HTMLInputElement).value).toMatchSnapshot()
  })

  it('Handles file input change', async () => {
    const fileInput = wrapper.findComponent(FileInput).find('input[type="file"]')

    Object.defineProperty(fileInput.element, 'files', {
      value: mockDataTransfer.files
    })

    await fileInput.trigger('change')

    await nextTick()

    const fileItem = wrapper.findAll('span').find((span) => span.text() === file.name) as DOMWrapper<HTMLSpanElement>

    expect(fileItem.exists()).toMatchSnapshot()
    expect(fileItem.text()).toMatchSnapshot()
  })

  it('Deletes file correctly', async () => {
    const fileInput = wrapper.findComponent(FileInput).find('input[type="file"]')

    Object.defineProperty(fileInput.element, 'files', {
      value: mockDataTransfer.files
    })

    await fileInput.trigger('change')

    await nextTick()

    const deleteButton = wrapper
      .findAllComponents(UiButton)
      .find((button) => button.attributes().icon === 'pi pi-times')

    await deleteButton?.trigger('click')

    await nextTick()

    const fileList = wrapper.find('ul')

    expect(fileList.exists()).toMatchSnapshot()
  })

  it('Equal button has type of submit when files is defined', async () => {
    const fileInput = wrapper.findComponent(FileInput).find('input[type="file"]')

    Object.defineProperty(fileInput.element, 'files', {
      value: mockDataTransfer.files
    })

    await fileInput.trigger('change')

    expect(findButtonByText(EQUAL)?.attributes().type).toMatchSnapshot()
  })

  it('Equal button reset type to button when clear button clicked', async () => {
    const fileInput = wrapper.findComponent(FileInput).find('input[type="file"]')

    Object.defineProperty(fileInput.element, 'files', {
      value: mockDataTransfer.files
    })

    await fileInput.trigger('change')

    expect(findButtonByText(EQUAL)?.attributes().type).toMatchSnapshot()

    const deleteButton = wrapper
      .findAllComponents(UiButton)
      .find((button) => button.attributes().icon === 'pi pi-times')

    await deleteButton?.trigger('click')

    expect(findButtonByText(EQUAL)?.attributes().type).toMatchSnapshot()
  })
})
