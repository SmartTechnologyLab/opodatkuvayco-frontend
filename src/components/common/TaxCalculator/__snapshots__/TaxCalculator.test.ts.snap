// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaxCalculator component > Calctulating the difference correctly 1`] = `"-1"`;

exports[`TaxCalculator component > Calctulating the division correctly 1`] = `"1"`;

exports[`TaxCalculator component > Calctulating the multiplication correctly 1`] = `"6"`;

exports[`TaxCalculator component > Calctulating the sum correctly 1`] = `"3"`;

exports[`TaxCalculator component > Deletes file correctly 1`] = `false`;

exports[`TaxCalculator component > Equal button has type of submit when files is defined 1`] = `"submit"`;

exports[`TaxCalculator component > Equal button reset type to button when clear button clicked 1`] = `"submit"`;

exports[`TaxCalculator component > Equal button reset type to button when clear button clicked 2`] = `"button"`;

exports[`TaxCalculator component > Handles file input change 1`] = `true`;

exports[`TaxCalculator component > Handles file input change 2`] = `"example.json"`;

exports[`TaxCalculator component > Set the number to negative correctly 1`] = `"0"`;

exports[`TaxCalculator component > When result is null or isCalculating is false - FileInput renders 1`] = `true`;

exports[`TaxCalculator component > When result is null or isCalculating is false - FileInput renders 2`] = `false`;

exports[`TaxCalculator component > When the button with number is clicked, the result input is mounted into DOM 1`] = `false`;

exports[`TaxCalculator component > When the button with number is clicked, the result input is mounted into DOM 2`] = `true`;

exports[`TaxCalculator component > When the reset button clicked, the result input is unmounted from DOM and the FileInput renders 1`] = `true`;

exports[`TaxCalculator component > When the reset button clicked, the result input is unmounted from DOM and the FileInput renders 2`] = `false`;
